### PR TITLE
[OneEmbedding] optimize unsorted_segment_sum when col is odd

### PR DESCRIPTION
**Pull request by guo-ran**

### CODE REVIEW

File: oneflow/user/kernels/data_shuffle_kernel.cu [hunk 1]
```diff
@@ -1099,6 +1101,133 @@ void ShuffleEmbeddingsGrad(cudaStream_t cuda_stream, ncclComm_t comm, int64_t pa
...
+template<typename K, typename IDX>
+__global__ void UnsortedSegmentHalfGpu(const IDX in_h2_elem_cnt, const IDX h2_inner_dim_size,
+                                       const IDX inner_dim_size, const half* data,
+                                       const K* segment_ids, const IDX num_segments,
+                                       half2* out_h2) {
+  CUDA_1D_KERNEL_LOOP_T(IDX, i, in_h2_elem_cnt) {
+    const IDX segment_id_idx = i / h2_inner_dim_size;
+    const IDX h2_inner_idx = i - segment_id_idx * h2_inner_dim_size;
+    const IDX inner_idx_0 = 2 * h2_inner_idx;
+    const IDX inner_idx_1 = 2 * h2_inner_idx + 1;
```
Comment (liujuncheng) [inline oneflow/user/kernels/data_shuffle_kernel.cu:1113]:
```suggestion
    const IDX inner_idx_1 = 2 * h2_inner_idx + 1;
```
```suggestion
    const IDX inner_idx_1 = inner_idx_0 + 1;
```

Comment (liujuncheng) [inline oneflow/user/kernels/data_shuffle_kernel.cu:1113]:
```suggestion
    const IDX inner_idx_1 = inner_idx_0 + 1;
```

File: oneflow/user/kernels/data_shuffle_kernel.cu [hunk 2]
```diff
@@ -1099,6 +1101,133 @@ void ShuffleEmbeddingsGrad(cudaStream_t cuda_stream, ncclComm_t comm, int64_t pa
...
+    const half* data_row = data + segment_id_idx * inner_dim_size;
+    half2 val;
+    val.x = data_row[inner_idx_0];
+    val.y = (inner_idx_1 >= inner_dim_size) ? static_cast<half>(0) : data_row[inner_idx_1];
+    const K origin_idx = segment_ids[segment_id_idx];
+    const IDX idx = origin_idx;
```
Comment (liujuncheng) [inline oneflow/user/kernels/data_shuffle_kernel.cu:1119]: 放到一行

Comment (liujuncheng) [inline oneflow/user/kernels/data_shuffle_kernel.cu:1119]: 放到一行

File: oneflow/user/kernels/data_shuffle_kernel.cu [hunk 3]
```diff
@@ -1099,6 +1101,133 @@ void ShuffleEmbeddingsGrad(cudaStream_t cuda_stream, ncclComm_t comm, int64_t pa
...
+struct UnsortedSegmentSumDiffInnerSize {
+  void operator()(ep::Stream* stream, const K* segment_ids, const T* data, int64_t num_segment_ids,
+                  int64_t num_segments, int64_t in_inner_dim_size, int64_t out_inner_dim_size,
+                  T* out) const {
+    UNIMPLEMENTED();
+  }
+};
+
+template<typename K>
+struct UnsortedSegmentSumDiffInnerSize<half, K> {
```
Comment (liujuncheng) [inline oneflow/user/kernels/data_shuffle_kernel.cu:1135]: UnsortedSegmentSumPad

Comment (liujuncheng) [inline oneflow/user/kernels/data_shuffle_kernel.cu:1135]: UnsortedSegmentSumPad

**Review by liujuncheng — APPROVED** [review]

**Review by MARD1NO — APPROVED** [review]

**Pull request merged**